com.unity.perception/Runtime/GroundTruth/Labelers/Visualization/OverlayPanel.cs

```diff
@@ -0,0 +1,231 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using UnityEngine.UI;
+
+namespace UnityEngine.Perception.GroundTruth
+{
+    /// <summary>
+    /// Interface that should be defined by a class that wants to be able to provide a image to the overlay
+    /// panel.
+    /// </summary>
+    public interface IOverlayPanelProvider
+    {
+        /// <summary>
+        /// Provides an image to the overlay panel.
+        /// </summary>
+        /// <returns>The image for the overlay panel.</returns>
+        Texture GetOverlayImage();
+    }
+
+    /// <summary>
+    /// Some labeler's result in a full screen image per frame. The overlay panel controls which of these labeler's image
+    /// is currently shown.
+    /// </summary>
+    public class OverlayPanel : MonoBehaviour
+    {
+        bool m_Enabled;
+        Dictionary<string, IOverlayPanelProvider> m_OverlayProviders = new Dictionary<string, IOverlayPanelProvider>();
+
+        GUIStyle m_LabelStyle;
+        GUIStyle m_SliderStyle;
+        GUIStyle m_SelectorToggleStyle;
+        GUIStyle m_WindowStyle;
+
+        float m_SegmentTransparency = 0.8f;
+        float m_BackgroundTransparency;
+
+        GameObject m_SegCanvas;
+        GameObject m_SegVisual;
+        RawImage m_OverlayImage;
+
+        int m_CachedHeight;
+        int m_CachedWidth;
+
+        bool m_ShowPopup = false;
+        Texture2D m_NormalDropDownTexture;
+        Texture2D m_HoverDropDownTexture;
+
+        /// <summary>
+        /// Registers a labeler as an overlay provider.
+        /// </summary>
+        /// <param name="provider">The overlay provider to register.</param>
+        /// <typeparam name="T">The type of the labeler, which must be an overlay provider and camera labeler</typeparam>
+        /// <returns>Did the labeler register properly</returns>
+        public bool RegisterOverlayProvider<T>(T provider) where T : CameraLabeler, IOverlayPanelProvider
+        {
+            var label = provider.GetType().Name;
+
+            m_OverlayProviders.Add(label, provider);
```

Comment:
Kicked this can down the road by getting rid of the need to register. Now is this still an issue, yeah, because of the drop down to choose which results that you want shown. I'm thinking of three solutions off of the top of my head:
1) tack a guid on the back of a name, but this is not good from a GUI perspective, both because the names will not make sense to the user, and the name will overrun the box width (which is already a problem)
2) tack a guid on the back of name, only if we have two labelers of the same name, also not great for reasons above, and also because it is going to put some special case code in the algorithm that we have to deal with.
3) I think the best solution, add a field to the labeler editor UI for a name, auto populate it with the name of the labeler class so a user doesn't have to add it themselves, but can if they want to for clarity. Auto tack on a 2.3,n if labelers of that type already exist.

I think if we choose 3, that we should spin off a separate task to do this.

Reply:
Makes sense to me. Definitely liking 3 the best, but I agree we can kick it down the road a bit. The new system is great.